pkg/registry/generic/registry/store.go

```diff
@@ -387,6 +440,39 @@ var (
 	errDeleteNow       = fmt.Errorf("delete now")
 )
 
+// return if we need to update the finalizers of the object
+func shouldUpdateFinalizers(accessor meta.Object, options *api.DeleteOptions) bool {
+	shouldOrphan := options != nil && options.OrphanDependents != nil && *options.OrphanDependents
+	orphanSet := false
+	for _, f := range accessor.GetFinalizers() {
+		if f == garbagecollector.OrphanFinalizerID {
+			orphanSet = true
+		}
+	}
+	return shouldOrphan != orphanSet
+}
+
+// add/remove the finalizers as the `options` requires
+func updateFinalizers(accessor meta.Object, options *api.DeleteOptions) {
```

Comment:
I still think you should take the desired list of finalizers and not compute that multiple places.